plugin/pkg/shoot/validator/admission.go

```diff
@@ -219,77 +271,79 @@ func (v *ValidateShoot) Admit(ctx context.Context, a admission.Attributes, o adm
-	}
+	return nil
+}
 
-	// We don't allow shoot to be created on the seed which is already marked to be deleted.
-	if seed != nil && seed.DeletionTimestamp != nil && a.GetOperation() == admission.Create {
-		return admission.NewForbidden(a, fmt.Errorf("cannot create shoot '%s' on seed '%s' already marked for deletion", shoot.Name, seed.Name))
-	}
+func (c *validationContext) validateScheduling(a admission.Attributes, shootLister corelisters.ShootLister, seedLister corelisters.SeedLister) error {
+	var (
+		shootIsBeingScheduled          = c.oldShoot.Spec.SeedName == nil && c.shoot.Spec.SeedName != nil
+		shootIsBeingRescheduled        = c.oldShoot.Spec.SeedName != nil && c.shoot.Spec.SeedName != nil && *c.shoot.Spec.SeedName != *c.oldShoot.Spec.SeedName
+		mustCheckSchedulingConstraints = shootIsBeingScheduled || shootIsBeingRescheduled
+	)
 
-	if oldShoot.Spec.SeedName != nil && !apiequality.Semantic.DeepEqual(shoot.Spec.SeedName, oldShoot.Spec.SeedName) &&
-		seed != nil {
-		if seed.Spec.Backup == nil {
-			return admission.NewForbidden(a, fmt.Errorf("cannot change seed name, because seed backup is not configured, for shoot %q", shoot.Name))
+	if mustCheckSchedulingConstraints {
+		if c.seed.DeletionTimestamp != nil {
+			return admission.NewForbidden(a, fmt.Errorf("cannot schedule shoot '%s' on seed '%s' that is already marked for deletion", c.shoot.Name, c.seed.Name))
 		}
 
-		oldSeed, err := v.seedLister.Get(*oldShoot.Spec.SeedName)
-		if err != nil {
-			return apierrors.NewBadRequest(fmt.Sprintf("could not find referenced seed: %+v", err.Error()))
+		if !helper.TaintsAreTolerated(c.seed.Spec.Taints, c.shoot.Spec.Tolerations) {
+			return admission.NewForbidden(a, fmt.Errorf("forbidden to use a seeds whose taints are not tolerated by the shoot"))
 		}
-		if oldSeed.Spec.Provider.Type != seed.Spec.Provider.Type {
-			return admission.NewForbidden(a, fmt.Errorf("cannot change seed name, because cloud provider for new seed (%s) is not equal to cloud provider for old seed (%s)", seed.Spec.Provider.Type, oldSeed.Spec.Provider.Type))
+
+		if allocatableShoots, ok := c.seed.Status.Allocatable[core.ResourceShoots]; ok {
+			scheduledShoots, err := getNumberOfShootsOnSeed(shootLister, c.seed.Name)
+			if err != nil {
+				return apierrors.NewInternalError(err)
+			}
+
+			if scheduledShoots >= allocatableShoots.Value() {
+				return admission.NewForbidden(a, fmt.Errorf("cannot schedule shoot '%s' on seed '%s' that already has the maximum number of shoots scheduled on it (%d)", c.shoot.Name, c.seed.Name, allocatableShoots.Value()))
+			}
 		}
 	}
 
-	if shoot.Spec.Provider.Type != cloudProfile.Spec.Type {
-		return apierrors.NewBadRequest(fmt.Sprintf("cloud provider in shoot (%s) is not equal to cloud provider in profile (%s)", shoot.Spec.Provider.Type, cloudProfile.Spec.Type))
-	}
+	if shootIsBeingRescheduled {
+		// TODO: this should be checked on both old and new Seed?
```

Comment:
@plkokanov @kris94 WDYT?

Reply:
It makes sense to check if the backup is configured in both seeds. Otherwise, if the old seed doesn't have backup enabled, but the new one does the admission would allow the migration, but there would be nowhere to get the ETCD data from and transfer it to the new seed.

Reply:
Ok thanks, added a check accordingly 👍